Crier documentation (#2431)
kyma-project/test-infra: docs/prow/crier.md

# Crier

Crier reports ProwJob status changes. For now it is responsible for Slack notifications as Plank is still reporting ProwJob statuses to the Github.

## Available reporters

> Crier supports multiple reporters, each reporter will become a crier controller. Reporters that can can be used:

Reviewer: `Reporters that can can be used:` -> `Reporters that can be used:`

    report_template: "Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <{{.Status.URL}}|View logs>"

  # "org/repo" slack config
  istio/proxy:
    job_types_to_report:
      - presubmit
    job_states_to_report:
      - error
    channel: istio-proxy-channel

  # "org" slack config
  istio:
    job_types_to_report:
      - periodic
    job_states_to_report:
      - failure
    channel: istio-channel
```

The Slack `channel` can be overridden at the ProwJob level via the `reporter_config.slack.channel` field:
```yaml
postsubmits:
  some-org/some-repo:
    - name: example-job
      decorate: true
      reporter_config:
        slack:
          channel: 'override-channel-name'
      spec:
        containers:
          - image: alpine
            command:
              - echo
```

## Current Slack notification settings

Crier does not send Slack notifications at all for presubmit jobs.

Reporter config:
```
job_types_to_report:
  - postsubmit
  - periodic
  - batch
```

If you don't want to report postsubmit or periodic jobs to report to Slack channel use `skip_report:true`.
If the job is still in testing phase we can set `optional: true`.

> ## Migration from plank for github report

Reviewer 2: ```suggestion
## Migration from Plank for GitHub report
```

> First, you need to disable GitHub reporting in Plank, add the `--skip-report=true` flag to the Plank [deployment](https://github.com/kyma-project/test-infra/blob/master/prow/cluster/components/11-plank_deployment.yaml).

Reviewer 2: ```suggestion
First, you need to disable GitHub reporting in Plank. To do that, add the `--skip-report=true` flag to the Plank [deployment](https://github.com/kyma-project/test-infra/blob/master/prow/cluster/components/11-plank_deployment.yaml).
```

> Before migrating, be sure plank is setting the [PrevReportStates field](https://github.com/kubernetes/test-infra/blob/de3775a7480fe0a724baacf24a87cbf058cd9fd5/prow/apis/prowjobs/v1/types.go#L566)

Reviewer 2: ```suggestion
Before migrating, be sure Plank is setting the [**PrevReportStates**](https://github.com/kubernetes/test-infra/blob/de3775a7480fe0a724baacf24a87cbf058cd9fd5/prow/apis/prowjobs/v1/types.go#L566) field
```

> by describing a finished presubmit ProwJob. Plank started to set this field after commit [2118178](https://github.com/kubernetes/test-infra/pull/10975/commits/211817826fc3c4f3315a02e46f3d6aa35573d22f), if not, you want to upgrade your plank to a version includes this commit before moving forward.

Reviewer 2: ```suggestion
by describing a finished presubmit ProwJob. Plank started to set this field after commit [`2118178`](https://github.com/kubernetes/test-infra/pull/10975/commits/211817826fc3c4f3315a02e46f3d6aa35573d22f). If it is not being set, you must upgrade your Plank to a version that includes this commit before moving forward.
```
Also, is this only for presubmit ProwJobs?

Author: No, it refers to the Plank in general.

> by describing a finished presubmit ProwJob. Plank started to set this field after commit [2118178](https://github.com/kubernetes/test-infra/pull/10975/commits/211817826fc3c4f3315a02e46f3d6aa35573d22f), if not, you want to upgrade your plank to a version includes this commit before moving forward.

Reviewer 2: Why is this sentence split into two lines? Can we make it one?

Flags required by Crier:

> - Point `config-path` and `--job-config-path` to your prow config and job configs accordingly.

Reviewer 2: ```suggestion
- Point `config-path` and `--job-config-path` to your Prow config and job configs accordingly.
```

> - Set `--github-worker` to be number of parallel github reporting threads you need.

Reviewer 2: ```suggestion
- Set `--github-worker` to be the number of parallel GitHub reporting threads you need.
```

> - Point `--github-endpoint` to ghproxy, if you have set that for plank.

Reviewer 2: ```suggestion
- Point `--github-endpoint` to ghproxy, if you have set that for Plank.
```

> - Bind github oauth token as a secret and set `--github-token-path` if you've have that set for plank.

Reviewer 2: ```suggestion
- Bind GitHub OAuth token as a Secret and set `--github-token-path` if you have that set for Plank.
```

> In your plank deployment, you can:

Reviewer 2: ```suggestion
In your Plank deployment, you can:
```

> In your plank deployment, you can:

Reviewer 2: `can` or `must`? Is this optional?

Author: To be honest I'm not sure if it is required but for sure it is recommended. 
I think it would be safer to use "must".

Author: It is required, I will change it to `must`.

- Remove the `--github-endpoint` flag.

> - Remove the github oauth secret, and `--github-token-path` flag if set.

Reviewer 2: ```suggestion
- Remove the GitHub OAuth Secret and the `--github-token-path` flag if set.
```

> - Add`--skip-report`, so plank will skip the reporting logic.

Reviewer 2: ```suggestion
- Add`--skip-report`, so Plank will skip the reporting logic.
```

> Both change should be deployed at the same time, if have an order preference, deploy crier first since report twice should just be a no-op.

Reviewer 2: ```suggestion
Both changes should be deployed at the same time. If you need to deploy them sequentially, deploy Crier first to avoid double-reporting.
```
This sentence is very unclear. I'm not sure if my suggestion reflects its original meaning. If not, let's discuss it.

## Available reporters

Crier supports multiple reporters. Each reporter will become a Crier controller. Reporters that can be used:
- GitHub reporter
- Slack reporter
- PubSub reporter

For any reporter that you want to use, you must mount your Prow configs and specify the `--config-path` and `--job-config-path` flags.

### GitHub reporter

You can enable the GitHub reporter in Crier by specifying the `--github-workers=N` flag.

You must also mount a GitHub OAuth token by specifying the `--github-token-path` flag, which defaults to `/etc/github/oauth`.

If you have a ghproxy deployed, also remember to point `--github-endpoint` to your ghproxy to avoid token throttle.

### Slack reporter

> **NOTE:** When you enable Crier for the first time, it will sent to Slack all ProwJobs matching the configured filtering criteria.

You can enable the Slack reporter in Crier by specifying the `--slack-workers` and `--slack-token-file` flags.

The `--slack-token-file` flag takes the path to the file containing the Slack [**OAuth Access Token**](https://api.slack.com/docs/oauth).

The **OAuth Access Token** can be obtained as follows:

> 1. Navigate to `https://api.slack.com/apps`.

Reviewer 2: ```suggestion
1. Navigate to [`https://api.slack.com/apps`](https://api.slack.com/apps).
```

1. Click **Create New App**.

> 1. Provide the **App Name** (e.g. *Prow Slack Reporter*) and **Development Slack Workspace** (e.g. *Kubernetes*).

Reviewer 2: ```suggestion
1. Provide the **App Name** (e.g. `Prow Slack Reporter`) and **Development Slack Workspace** (e.g. `Kubernetes`).
```

1. Click **Permissions**.
1. Add the `chat:write.public` scope using the **Scopes / Bot Token Scopes** dropdown and click **Save Changes**.
1. Click **Install App to Workspace**.
1. Click **Allow** to authorize the OAuth scopes.
1. Copy the **OAuth Access Token**.

Once the access token is obtained, you can create a Secret in the cluster using that value:

```shell
kubectl create secret generic slack-token --from-literal=token="{ACCESS_TOKEN}"
```

> Furthermore, to make this token available to Crier, mount the `slack-token` Secret as a volume and set the `--slack-token-file` flag in the deployment spec.

Reviewer 2: ```suggestion
Furthermore, to make this token available to Crier, mount the `slack-token` Secret as a volume and set the `--slack-token-file` flag in the deployment spec:
```

    channel: my-slack-channel
    # The template shown below is the default
    report_template: "Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <{{.Status.URL}}|View logs>"

  # "org/repo" slack config
  istio/proxy:
    job_types_to_report:
      - presubmit
    job_states_to_report:
      - error
    channel: istio-proxy-channel

  # "org" slack config
  istio:
    job_types_to_report:
      - periodic
    job_states_to_report:
      - failure
    channel: istio-channel
```

The Slack channel can be overridden at the ProwJob level via the **reporter_config.slack.channel** field:

```yaml
postsubmits:
  some-org/some-repo:
    - name: example-job
      decorate: true
      reporter_config:
        slack:
          channel: 'override-channel-name'
      spec:
        containers:
          - image: alpine
            command:
              - echo
```

## Current Slack notification settings

Crier does not send any Slack notifications for presubmit jobs.

Reporter config:

```
job_types_to_report:
  - postsubmit
  - periodic
  - batch
```

If you don't want to configure postsubmit or periodic jobs to report to a Slack channel, use `skip_report:true`.

> If the job is still in the testing phase, we can set `optional: true`.

Reviewer 2: ```suggestion
If the job is still in the testing phase, you can set `optional: true`.
```

> ## Migration from Plank for GitHub report

Reviewer 2: ```suggestion
## Migration from Plank to GitHub report
```

First, you need to disable GitHub reporting in Plank. To do that, add the `--skip-report=true` flag to the Plank [deployment](https://github.com/kyma-project/test-infra/blob/master/prow/cluster/components/11-plank_deployment.yaml).

> Before migrating, upgrade your Plank to a version that includes this commit [`2118178`](https://github.com/kubernetes/test-infra/pull/10975/commits/211817826fc3c4f3315a02e46f3d6aa35573d22f).

Reviewer 2: ```suggestion
Before migrating, upgrade your Plank to a version that includes the commit [`2118178`](https://github.com/kubernetes/test-infra/pull/10975/commits/211817826fc3c4f3315a02e46f3d6aa35573d22f).
```

Flags required by Crier:
- Point `config-path` and `--job-config-path` to your Prow config and job configs accordingly.
- Set `--github-worker` to be the number of parallel GitHub reporting threads you need.
- Point `--github-endpoint` to ghproxy, if you have set that for Plank.
- Bind GitHub OAuth token as a Secret and set `--github-token-path` if you have that set for Plank.

In your Plank deployment, you must:
- Remove the `--github-endpoint` flag.
- Remove the GitHub OAuth Secret and the `--github-token-path` flag if set.

> - Add`--skip-report`, so Plank will skip the reporting logic.

Reviewer 2: ```suggestion
- Add `--skip-report`, so Plank will skip the reporting logic.
```
